### PR TITLE
Enable PluginData for users

### PR DESCRIPTION
**Pull request by gzigzigzeo**
For now, PluginData is attachable to AccessRequest only. While it is still possible to attach PluginData to a whole entity (without id specified), a plugin might not have access to AccessRequests at all (as event handler, for example). We need a way to store plugin-wide persistent data. The easiest approach is to attach plugin-wide data to a plugin primary user.

This PR enables attaching PluginData to User.

Please, check TODOs: 

1. lib/auth/auth_with_roles.go: I am not sure how to correctly perform auth checks in UpdatePluginData() and GetPluginData(). We must ensure not only that a user has access to "UserPluginData", but that a user has access to a specific user in scope (himself).
2. AccessRequests have expiration time. PluginData TTL is explicitly set to parent AccessRequest.Expires + 1h on create. This ensures that there would be no orphaned PluginData. Normally, user does not have any TTL. So, PluginData is created with no expiration by default. That might become a problem. However, I am not sure that this is the scope of this PR.

### CODE REVIEW

File: lib/services/local/dynamic_access.go
Comment (tcsc) [inline lib/services/local/dynamic_access.go:430]: Would...

```suggestion
			if (exp.IsZero()) {
```
... work here?

File: lib/services/local/dynamic_access.go
Comment (tcsc) [inline lib/services/local/dynamic_access.go:356]: Is there some sort of interfaced-based approach we could take to this? E.g. a `ExpiryGetter` interface that is implemented by all resource types that you need to get an expiry for?

**Edit:** 
Actually, that would
1. be overkill for now - the switch does the job fine, and
2. probably mean you'd have to pull the arbitrary resource out of the back end,  un-marshall it, query for the correct interface and only _then_ pull the expiry out. If the expiry just uses the backend record's expiry value (as it does in a full 50% of the covered cases), that would be a stupid amount of overwork. 

Leaving the comment here for the sake of discussion, but consider it resolved.

**Review by tcsc — COMMENTED** [review]
Looks good in general. I can't really answer the questions in your PR comment (well, _could_. They just wouldn't be very _good_ answers...), so I'll leave that to others.

File: lib/services/local/dynamic_access.go
Comment (gzigzigzeo) [reply lib/services/local/dynamic_access.go]: Absolutely. Thank you, fixed.

File: lib/services/local/dynamic_access.go
Comment (gzigzigzeo) [reply lib/services/local/dynamic_access.go]: Yes, there actually is. I did one more pass over the existing architecture. I think, I now understand it better. I have re-implemented `PluginData` from scratch as a separate service. Now, there is isolated `PluginData` service which is responsible for storage only, and there are `Server` methods which are responsible for high-level checks. Thank you for pointing to it.

**Comment by Joerger** [comment]
@gzigzigzeo Can you help me understand the use case for separating out plugin data? Is this for the event-handler login locking? 

1) The user doesn't need access to plugin data, only the plugin user right? In which case the plugin user's role will need an allow rule for `KindPluginData` and verbs `read,list,update,delete`.
2) That's a good point and probably should be solved in this Pr, not sure how yet

I'll take a deeper look at this on Monday

File: lib/services/local/dynamic_access.go
Comment (tcsc) [reply lib/services/local/dynamic_access.go]: Oops - forgto an `!`

File: lib/services/local/plugin_data.go
Comment (tcsc) [inline lib/services/local/plugin_data.go:101]: While this is being moved about, is it worth (or even a good idea...) changing this so that it keeps trying until the context expires, rather than some arbitrary number of retries?

**Review by tcsc — COMMENTED** [review]
LGTM - with the caveat that the permissions enforcement still needs review by someone with a deeper understanding of it.

File: lib/services/local/dynamic_access.go
Comment (gzigzigzeo) [reply lib/services/local/dynamic_access.go]: > @gzigzigzeo Can you help me understand the use case for separating out plugin data? Is this for the event-handler login locking?

Not exactly. User locking data is now in-memory. I do not see a lot of benefits of persisting it because `LockPeriod` is relatively short. Loosing it wouldn't hurt much. In the another hand, persisting it would complicate things a bit. 

The motivation behind this is the following:

1. Get rid of disk state in `event-handler`. For now, it is very compact and consists of a bunch of cursors (strings and ints). Disk turned out to be unreliable storage, especially in k8s. As far as I understood, clients are experiencing problems. Sometimes it gets corrupted and lost. Moving this state to the API would solve the reliability problems.
2. We are going to need to provide some kind of a simple persistent storage to our feature WASM plugin authors.
3. As far as I know, this would potentially be useful for other plugins we have (@marshall-lee might add his thoughts here).

I have separated `PluginData` because it was too tightly coupled to `AccessRequest` and I needed to expand it to work with users.

> 1. The user doesn't need access to plugin data, only the plugin user right? In which case the plugin user's role will need an allow rule for `KindPluginData` and verbs `read,list,update,delete`.

That's right. I think, that would be enough for now. Plugin user should have access to their own plugin data, admin user should be able to access a plugin data of any user. I think, ideally, we would want to make access configureable using labels.

> 2. That's a good point and probably should be solved in this Pr, not sure how yet

I think that for now we could introduce a simple workaround which sets `PluginData` TTL to Now()+1h when user gets deleted. 

But I also think that we need a universal way of managing resource TTLs and relations. The backend service seems the good candidate to put this responsibility on. However, I am not yet sure how exactly.

**Comment by gzigzigzeo** [comment]
@Joerger 

> @gzigzigzeo Can you help me understand the use case for separating out plugin data? Is this for the event-handler login locking?

Not exactly. User locking data is now in-memory. I do not see a lot of benefits of persisting it because `LockPeriod` is relatively short. Loosing it wouldn't hurt much. In the another hand, persisting it would complicate things a bit. 

The motivation behind this is the following:

1. Get rid of disk state in `event-handler`. For now, it is very compact and consists of a bunch of cursors (strings and ints). Disk turned out to be unreliable storage, especially in k8s. As far as I understood, clients are experiencing problems. Sometimes it gets corrupted and lost. Moving this state to the API would solve the reliability problems.
2. We are going to need to provide some kind of a simple persistent storage to our feature WASM plugin authors.
3. As far as I know, this would potentially be useful for other plugins we have (@marshall-lee might add his thoughts here).

I have separated `PluginData` because it was too tightly coupled to `AccessRequest` and I needed to expand it to work with users.

> 1. The user doesn't need access to plugin data, only the plugin user right? In which case the plugin user's role will need an allow rule for `KindPluginData` and verbs `read,list,update,delete`.

That's right. I think, that would be enough for now. Plugin user should have access to their own plugin data, admin user should be able to access a plugin data of any user. I think, ideally, we would want to make access configureable using labels.

> 2. That's a good point and probably should be solved in this Pr, not sure how yet

I think that for now we could introduce a simple workaround which sets `PluginData` TTL to Now()+1h when user gets deleted. 

But I also think that we need a universal way of managing resource TTLs and relations. The backend service seems the good candidate to put this responsibility on. However, I am not yet sure how exactly.

File: lib/services/local/dynamic_access.go
Comment (gzigzigzeo) [reply lib/services/local/dynamic_access.go]: No worries, it has been moved to another file, and ! is in place there :-)

File: lib/services/local/plugin_data.go
Comment (marshall-lee) [reply lib/services/local/plugin_data.go:101]: If there's some kind of exponential backoff then it's okay to rely on a context expiration. But it's better to handle `CompareFailed` on a client side and do the backoff there (like we do in `teleport-plugins`).

So whenever you don't have a backoff, you should definitely limit the try count cause otherwise you get the backend DoS'ed before the context expiration.

Actually, the `maxCmpAttempts = 1` would be enough but in order to reduce the network round trips of a client & its backoff, there're some multiple tries on a backend side too.

**Comment by marshall-lee** [comment]
> Can you help me understand the use case for separating out plugin data? Is this for the event-handler login locking?

Plugin Data is already designed as resource-agnostic, it's not really coupled with `AccessRequest`. However, the only supported resource by now is `AccessRequest` — the backend refuses to write PluginData with any `SubKind` different from `"AccessRequest"`.

The desired use case, in short:

Generally, the `PluginData` resource can be thought as some  kind of extension of any other resource in Teleport. However, there're cases where you don't have a particular resources you can connect to. You just want a small persistent storage for plugin needs. And if we enable storing PluginData for users, then we can treat plugin user's PluginData as a plugin's storage.

File: lib/auth/auth.go
Comment (Joerger) [inline lib/auth/auth.go:3388]: Let's keep the plugin data kind check on the backend, see comment below.
```suggestion
func (a *Server) GetPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
	return a.Services.PluginData.GetPluginData(ctx, filter)
}

// GetPluginData loads all plugin data matching the supplied filter.
func (a *Server) UpdatePluginData(ctx context.Context, params types.PluginDataUpdateParams) error {
	// Load parent resource expiration value. Nil means that a parent resource does not exist.
	// Non-existent parent resource is fine on update. There are the cases when we need to access
	// PluginData of a just deleted resource. We need to prevent creation of PluginData entries
	// for a missing resources.
	//
	// Expires itself is needed to prevent orphaned plugin data, we automatically
	// configure new instances to expire shortly after the resource to which they are associated.
	expires, err := a.getResourceExpiry(ctx, params.Resource, params.Kind)
	if err != nil {
		return trace.Wrap(err)
	}

	return a.PluginData.UpsertPluginData(ctx, params, expires)
}
```

File: lib/services/local/plugin_data.go
Comment (Joerger) [inline lib/services/local/plugin_data.go:37]: Two changes: 
  1) No need to separate it into multiple functions since all resource kinds (for now) use the same logic.
  2) Let's move the filter kind check back to here and out of the server.
```suggestion
// GetPluginData loads all plugin data matching the supplied filter.
func (s *PluginDataService) GetPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
	if !isAllowedPluginDataType(filter.Kind) {
		return nil, trace.BadParameter("unsupported resource kind %q", filter.Kind)
	}
	
	// Filters which specify Resource are a special case since they will match exactly zero or one
```

File: lib/services/local/plugin_data.go
Comment (Joerger) [inline lib/services/local/plugin_data.go:214]:
```suggestion
func isAllowedPluginDataKind(kind string) bool {
    return apiutils.SliceContainsStr([]string{types.KindAccessRequest, types.KindUser}, kind)
}
```

File: lib/services/local/plugin_data.go
Comment (Joerger) [inline lib/services/local/plugin_data.go:92]: Same as above.
```suggestion
func (s *PluginDataService) UpsertPluginData(ctx context.Context, params types.PluginDataUpdateParams, parentExpiresAt *time.Time) error {
	if !isAllowedPluginDataType(filter.Kind) {
		return nil, trace.BadParameter("unsupported resource kind %q", filter.Kind)
	}
	
	retryPeriod := retryPeriodMs * time.Millisecond
```

File: lib/auth/auth_with_roles.go
Comment (Joerger) [inline lib/auth/auth_with_roles.go:1234]: > 1. lib/auth/auth_with_roles.go: I am not sure how to correctly perform auth checks in UpdatePluginData() and GetPluginData(). We must ensure not only that a user has access to "UserPluginData", but that a user has access to a specific user in scope (himself).

It should be fine as is. `plugin_data` is a special permission which gives permission to update plugin data on any resource, without requiring update permissions for the resource in question. See https://github.com/gravitational/teleport/pull/6210.
```suggestion
```

File: lib/auth/auth_with_roles.go
Comment (Joerger) [inline lib/auth/auth_with_roles.go:1257]:
```suggestion
```

File: lib/auth/auth_with_roles.go
Comment (Joerger) [inline lib/auth/auth_with_roles.go:1258]:
```suggestion
		if err := a.action(apidefaults.Namespace, types.KindUserPluginData, types.VerbUpdate, types.VerbCreate); err != nil {
```

File: lib/services/local/plugin_data.go
Comment (Joerger) [inline lib/services/local/plugin_data.go:183]: > 2. AccessRequests have expiration time. PluginData TTL is explicitly set to parent AccessRequest.Expires + 1h on create. This ensures that there would be no orphaned PluginData. Normally, user does not have any TTL. So, PluginData is created with no expiration by default. That might become a problem. However, I am not sure that this is the scope of this PR.

The way this has been handled in the past with other resources which could be orphaned once a user is deleted (such as user recovery tokens) is to make the resource key prefix have the parent resource key as a prefix. Ex: `/web/users/{user_id}`, `/web/users/{user_id}/recoverycodes/{code}`. This way, when the user is deleted, any resource with `web/user/{user_id}` as a key prefix is deleted as well.

```suggestion
// pluginDataKey constructs PluginData entry backend key
func pluginDataKey(kind string, name string) []byte {
	if kind == types.KindUser {
		return backend.Key(webPrefix, usersPrefix, name, pluginDataPrefix, paramsPrefix)
	}
	return backend.Key(pluginDataPrefix, kind, name, paramsPrefix)
}
```

**Comment by russjones** [comment]
@tcsc @Joerger @gzigzigzeo Let's wait for @fspmarshall to approve this PR before we merge.

@fspmarshall Do you mind doing a review?

**Review by fspmarshall — DISMISSED** [review]
PR looks good to me overall.  I would prefer some kind of default upper-limit for how long `PluginData` for non-expiring users is stored, even if it is a very long ttl (e.g. 1 year).  If injecting a very long default TTL is impractical for some reason, another workaround could be to allow plugins to specify how long they need the data to live (requiring the supplied ttl to be non-zero if the associated resource is non-expiring).  Fixing this now isn't a hard requirement, but I'd prefer it.

I do have a concern about the proposed usecase of storing the cursor from `event-handler`.  As currently implemented, `event-handler` updates the cursor for each event that is processed.  Porting this over 1-to-1 would result in far too much database I/O in my opinion.  Very large clusters can produce a lot of events, and we are already experiencing issues due to teleport producing too much database I/O.  Would it be acceptable to continue storing the cursor on disk, and have a periodic operation which syncs it to `PluginData` every few minutes?

File: lib/services/local/plugin_data.go
Comment (gzigzigzeo) [reply lib/services/local/plugin_data.go:214]: In my opinion, checking slice worths it when a target list is dynamic. Here we check essentially a constant value, so, it seems that a simple if would be much readable and fast. Are you sure we need `SliceContainsStr` here?

File: lib/services/local/plugin_data.go
Comment (gzigzigzeo) [reply lib/services/local/plugin_data.go:183]: I realised that the case I described above: when PluginData needs to be around for some time after a parent object gets destroyed is not be applicable to users. So, I replaced it here like you proposed. ^

**Comment by gzigzigzeo** [comment]
@fspmarshall 

> PR looks good to me overall. I would prefer some kind of default upper-limit for how long `PluginData` for non-expiring users is stored, even if it is a very long ttl (e.g. 1 year). If injecting a very long default TTL is impractical for some reason, another workaround could be to allow plugins to specify how long they need the data to live (requiring the supplied ttl to be non-zero if the associated resource is non-expiring). Fixing this now isn't a hard requirement, but I'd prefer it.

I agree that PluginData needs a TTL, but do not see a reliable way to set it for now. I think, that setting it to a long value would is not a good idea because it would lead to a silent destroy of potentially sensitive PluginData at some point. Hence, a TTL should be long - but how exactly long to distinguish it from infinite?

However, I just made a change which ensures that PluginData gets destroyed along with a user. 

> I do have a concern about the proposed usecase of storing the cursor from `event-handler`. As currently implemented, `event-handler` updates the cursor for each event that is processed. Porting this over 1-to-1 would result in far too much database I/O in my opinion. Very large clusters can produce a lot of events, and we are already experiencing issues due to teleport producing too much database I/O. Would it be acceptable to continue storing the cursor on disk, and have a periodic operation which syncs it to `PluginData` every few minutes?

I agree, PluginData can't be synced on every event as disk does now. There would be some kind of checkpoints: either periodical, or timed, or both. 

I think, we could do it either way: in case user wants to use disk storage, we use PluginData as the backup storage. If disk gets corrupted, we retrieve last known cursors from the API. Otherwise, we use just PluginData keeping in mind that in case of failure, some events could be duplicated.

File: lib/auth/auth.go
Comment (Joerger) [inline lib/auth/auth.go:3368]: nit: I don't think this comment is necessary, this is explained in detail on the backend.
```suggestion
```

**Review by Joerger — APPROVED** [review]

**Review by tcsc — APPROVED** [review]
Approved, once @Joerger's suggestions are applied

**Pull request closed without merging**
